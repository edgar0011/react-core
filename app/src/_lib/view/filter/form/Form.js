// @flow
import React, { PureComponent } from 'react'
import type { Node, Element, ComponentType } from 'react'
import type { HeadlessRender } from '../../common/HeadlessComponentBase'

export type FormComponents = {
  FormComponent: ComponentType<any>,
  FormFieldComponent: ComponentType<any>,
  ActionGroupComponent: ComponentType<any>
}

export type FlexLayout = {
  direction?: string,
  align?: string,
  self?: string,
  justify?: string,
  size?: number | string,
  padding?: number | string,
  margin?: number | string,
  wrap?: boolean,
  width?: string,
  minWidth?: string,
  maxWidth?: string
}

export type FlexFieldLayout = FlexLayout & {
  minWidth?: string
}

export type Field = {
  name: string,
  label?: string,
  Component: ComponentType<any>,
  value: any,
  valueDecorator?: (value: any, field: Field) => any,
  onChange?: (name: string, newValue: any) => void,
  render?: Function
}

export type Error = {
  code: string,
  message?: string
}

export type Errors = Array<Error> | any

export type RenderFieldProps = {
  field: Field,
  flexFieldLayout: FlexFieldLayout,
  fieldValue: any,
  components: FormComponents,
  fields: Array<Field>
}

export type FormProps = {
  fields: Array<Field>,
  errors?: Errors,
  flexFieldLayout: FlexFieldLayout,
  components: FormComponents,
  renderFieldMethod: (props: RenderFieldProps) => Array<Element<any>>
} & HeadlessRender

// aka FormBody
class Form extends PureComponent<FormProps> {
  props: FormProps

  render() {
    const {
      fields, flexFieldLayout, errors, render, RenderComponent, renderFieldMethod, components, ...formProps
    }: FormProps = this.props

    const hasErrors = !!(errors && (errors.length || Object.values(errors).length))

    const fieldNodes =
      fields.map(field => renderFieldMethod({ field, flexFieldLayout, fieldValue: field.value, components, fields }))

    const passProps = {
      fieldNodes,
      hasErrors,
      errors,
      components,
      ...formProps
    }

    let rendered: Node
    if (render) {
      rendered = render(passProps)
    } else {
      rendered = RenderComponent ? <RenderComponent {...passProps} /> : null
    }
    return rendered
  }
}

export default Form
