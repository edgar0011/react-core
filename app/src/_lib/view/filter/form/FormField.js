// @flow
import React, { PureComponent } from 'react'
import type { Node } from 'react'
import type { HeadlessRender } from '../../common/HeadlessComponentBase'

type FormFieldProps = {

} & HeadlessRender

class FormField extends PureComponent<FormFieldProps> {

  props: FormFieldProps

  render() {
    const { render, RenderComponent, ...props }: FormFieldProps = this.props

    let rendered: Node
    if (render) {
      rendered = render(props)
    } else {
      rendered = RenderComponent ? <RenderComponent {...props} /> : null
    }
    return rendered
  }
}

export default FormField
