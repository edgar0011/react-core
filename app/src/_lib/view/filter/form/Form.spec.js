import React from 'react'
import Form from './Form'
import FormField from './FormField'

const renderFieldMethod = (field) => <FormField {...field} />

describe('Form', () => {
  it('should render Form', () => {
    const fields = [
      {
        name: 'input1',
        label: 'Input1'
      }
    ]
    const fieldValues = {
      input1: 'Hello Form'
    }

    global.shallowSnapshot(<Form fields={fields} fieldValues={fieldValues} renderFieldMethod={renderFieldMethod} />)
  })
})
