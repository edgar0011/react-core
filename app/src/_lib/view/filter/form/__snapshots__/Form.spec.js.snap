// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render Form 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form
    fieldValues={
      Object {
        "input1": "Hello Form",
      }
    }
    fields={
      Array [
        Object {
          "label": "Input1",
          "name": "input1",
        },
      ]
    }
    renderFieldMethod={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
